Add zlib tests and fix runtime bug

mpg opened the pull request:
## Description

This PR fixes a gap in our testing: we didn't test any build with zlib (for TLS record compression) enabled. This is a deprecated option but it should nonetheless be tested until it's removed for real.

The usefulness of these tests is illustrated by the fact that in their absence, we introduced two bug in 2.19 that caused us to fail to build or run with zlib enabled: #2859 plus another one that was found by the tests and is fixed in this PR.

__Dependencies:__ This PR is based on #2963 which fixes a bug in building with zlib - the added test obviously wouldn't pass without that fix.

## Status
**READY**

## Requires Backporting

Yes - for the tests (the bugfix is only applicable to development, the LTS branches are free from the bug as the tests show).

- [x] Mbed TLS 2.16 https://github.com/ARMmbed/mbedtls/pull/2972
- [x] Mbed TLS 2.7 https://github.com/ARMmbed/mbedtls/pull/2976

## Migrations

NO

## Additional comments

No ChangeLog entry as this is only a change in tests, not in the library, and the runtime bug probably doesn't need it own entry as it was hidden by the compile bug.

## Steps to test or reproduce

Run `tests/scripts/all.sh 'test_zlib*'` and observe the result, in particular the two lines with `Default (compression enabled)` in the output of `ssl-opt.sh`.

mpg commented:
The CI's pr-merge job succeeds, and the pr-head jobs only fails the `check-python-files` step, which is independent from this PR and was fixed in development in the meantime. This PR is based on #2963 which is a month old, rather than on current development.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -697,6 +697,31 @@ component_test_full_cmake_gcc_asan () {
+component_test_zlib_make() {
+    msg "build: zlib enabled, make"
+    scripts/config.py set MBEDTLS_ZLIB_SUPPORT
+    make ZLIB=1 CFLAGS='-Werror -O1'
```
This adds a build-time requirement that the development files for zlib must be available (`zlib.h` and `libz.so`). Should we assume that these files are always available? If so we should add a function similar to `check_tools` to check their availability. If not we should add a function `support_test_zlib_make` that checks their availability (and also `test_zlib_cmake`). Perhaps something like (untested):

```
support_test_zlib_make () {
    ret=0
    cat <<'EOF' >support_test_zlib_make_$$.c
#include "zlib.h"
int main(void) { return 0; }
EOF
    gcc -o support_test_zlib_make_$$.exe support_test_zlib_make_$$.c -lz || ret=$?
    rm -f support_test_zlib_make_$$.*
    return $ret
}
```
and
```
support_test_zlib_make () {
    support_test_zlib_cmake "$@"
}
```

mpg commented:
@gilles-peskine-arm Thanks for your feedback! I added `support_` functions based on your suggestion. 
I also rebased on #2963 since it was itself rebased on development to fix CI issues and conflicts in the ChangeLog. This is ready for you to review again.

mpg commented:
Note: I'll update the backports once this PR reaches an approved state.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -697,6 +697,45 @@ component_test_full_cmake_gcc_asan () {
+int main(void) { return 0; }
+EOF
+    gcc -o ${base}.exe ${base}.c -lz 2>/dev/null
+    ret=$?
```
`ret=$?` relies on `set -e` being inactive. Given that `support_test_zlib_make` is only ever called through
```
if ! support_$component; then …
```
this works, because `set -e` is deactivated in the `if` clause. But it's fragile. That's why I suggested a more complex version whose behavior is independent of whether `set -e` is in effect.
